server/src/main/java/io/druid/segment/loading/LocalDataSegmentPuller.java

```diff
@@ -17,58 +17,196 @@
+      final File[] files = sourceFile.listFiles();
+      if (files == null) {
+        throw new SegmentLoadingException("No files found in [%s]", sourceFile.getAbsolutePath());
       }
+      final FileUtils.FileCopyResult result = new FileUtils.FileCopyResult(sourceFile);
+      for (final File oldFile : files) {
+        if (oldFile.isDirectory()) {
+          log.info("[%s] is a child directory, skipping", oldFile.getAbsolutePath());
+          continue;
+        }
 
-      log.info("Unzipping local file[%s] to [%s]", path, dir);
+        result.addFiles(
+            FileUtils.retryCopy(
+                Files.asByteSource(oldFile),
+                new File(dir, oldFile.getName()),
+                shouldRetryPredicate(),
+                10
+            ).getFiles()
+        );
+      }
+      log.info("Coppied %d bytes from [%s] to [%s]", result.size(), sourceFile.getAbsolutePath(), dir.getAbsolutePath());
+      return result;
+    }
+    if (CompressionUtils.isZip(sourceFile.getName())) {
       try {
-        CompressionUtils.unzip(path, dir);
+        final FileUtils.FileCopyResult result = CompressionUtils.unzip(
+            Files.asByteSource(sourceFile),
+            dir,
+            shouldRetryPredicate(),
+            false
+        );
+        log.info("Unzipped %d bytes from [%s] to [%s]", result.size(), sourceFile.getAbsolutePath(), dir.getAbsolutePath());
+        return result;
       }
       catch (IOException e) {
-        throw new SegmentLoadingException(e, "Unable to unzip file[%s]", path);
+        throw new SegmentLoadingException(e, "Unable to unzip file [%s]", sourceFile.getAbsolutePath());
       }
     }
+    if (CompressionUtils.isGz(sourceFile.getName())) {
+      final File outFile = new File(dir, CompressionUtils.getGzBaseName(sourceFile.getName()));
+      final FileUtils.FileCopyResult result = CompressionUtils.gunzip(
+          Files.asByteSource(sourceFile),
+          outFile,
+          shouldRetryPredicate()
+      );
+      log.info(
+          "Gunzipped %d bytes from [%s] to [%s]",
+          result.size(),
+          sourceFile.getAbsolutePath(),
+          outFile.getAbsolutePath()
+      );
+      return result;
+    }
+    throw new SegmentLoadingException("Do not know how to handle source [%s]", sourceFile.getAbsolutePath());
+  }
+
+
+  @Override
+  public InputStream getInputStream(URI uri) throws IOException
+  {
+    return buildFileObject(uri).openInputStream();
+  }
+
+  /**
+   * Returns the "version" (aka last modified timestamp) of the URI of interest
+   * @param uri The URI to check the last modified timestamp
+   * @return The last modified timestamp of the URI in String format
+   */
+  @Override
+  public String getVersion(URI uri)
+  {
+    return String.format("%d", buildFileObject(uri).getLastModified());
+  }
+
+  @Override
+  public Predicate<Throwable> shouldRetryPredicate()
+  {
+    // TODO: smarter retries
```

Comment:
can we leave a comment here about why smarter retries are necessary and remove the TODO?